cpu/sam0_common/periph/adc.c

```diff
@@ -121,23 +165,41 @@ int adc_sample(adc_t line, adc_res_t res)
         case ADC_RES_12BIT:
             resolution = ADC_CTRLB_RESSEL_12BIT;
             break;
+#else
+    case ADC_RES_8BIT:
+        resolution = ADC_CTRLC_RESSEL_8BIT;
+        break;
+    case ADC_RES_10BIT:
+        resolution = ADC_CTRLC_RESSEL_10BIT;
+        break;
+    case ADC_RES_12BIT:
+        resolution = ADC_CTRLC_RESSEL_12BIT;
+        break;
+#endif
         default:
             resolution = 0xFFFFFFFF;
+            DEBUG("adc: Resolution arg not applicable\n");
             return -1;
     }
```

Comment:
Actually, this switch statement can be fully eliminated: we can just say
```c
_adc_configure(res);
```
insted, all we need to do is to override the `ADC_RES_xBIT` enum values for each cpu. For example for the `samd21`  (`samd21/include/cpu_conf.h`):
```c
/**
 * @brief   Override the ADC resolution configuration
 * @{
 */
#define HAVE_ADC_RES_T
typedef enum {
    ADC_RES_6BIT  = 0xff,     /**< not supported */
    ADC_RES_8BIT  = ADC_CTRLC_RESSEL_8BIT,     /**< ADC resolution: 8 bit */
    ADC_RES_10BIT = ADC_CTRLB_RESSEL_10BIT,     /**< ADC resolution: 10 bit */
    ADC_RES_12BIT = ADC_CTRLB_RESSEL_12BIT,     /**< ADC resolution: 12 bit */
    ADC_RES_14BIT = 0xfe,              /**< not supported */
    ADC_RES_16BIT = ADC_CTRLB_RESSEL_16BIT_Val               /**< ADC resolution: 16 bit */
} adc_res_t;
/** @} */
```

The validity of this parameter can further be asserted in the beginning of the function with a simple
```c
assert(res <= 0x3);
```